torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py

```diff
@@ -1,76 +1,328 @@
+    Handles ``__torch_function__`` dispatch for the sharded_masked_fill op.
+    We first narrow down the mask to the size of local tensor if the mask
+    contains the sharding dim and then apply the mask to the local tensor.
+
+    Args: same as ``torch.Tensor.masked_fill``.
+
+    Return:
+        local_tensor (Tensor): New local tensor to build the sharded tensor.
+        sharding_spec (:class:`torch.distributed._shard.sharding_spec.ShardingSpec`):
+            sharding spec of the new sharded tensor.
+        new_st_size (torch.Size): Size of the new sharded tensor.
     """
-    input = args[0]
-    p = kwargs.get("p")
+    st = args[0]
+    mask = args[1]
+    value = args[2]
+    current_rank = dist.get_rank(pg)  # type: ignore[attr-defined]
+    sharding_dim = st.sharding_spec().dim  # type: ignore[attr-defined]
+    narrow_idx = None
+    for idx in range(-1, -mask.dim() - 1, -1):
+        if st.dim() + idx == sharding_dim:
+            narrow_idx = idx
+    if narrow_idx is not None:
+        rank_idx = None
+        for idx, placement in enumerate(st._sharding_spec.placements):  # type: ignore[attr-defined]
+            if placement.rank() == current_rank:  # type: ignore[index]
+                rank_idx = idx  # type: ignore[attr-defined]
+        shard_metadata = st.metadata().shards_metadata[rank_idx]  # type: ignore[index]
+        mask = mask.narrow(
+            narrow_idx,
+            shard_metadata.shard_offsets[sharding_dim],
+            shard_metadata.shard_sizes[sharding_dim],
+        )
+    local_tensor = st.local_tensor().masked_fill(mask, value)
+    return local_tensor, st.sharding_spec(), st.size()
+
 
-    # Validate types
-    if not isinstance(input, ShardedTensor):
-        raise TypeError("input needs to be a ShardedTensor")
-    local_tensor = input.local_tensor()
-    local_tensor = torch.nn.functional.dropout(local_tensor, p=p)
-    return ShardedTensor._init_from_local_tensor(
-        local_tensor.contiguous(),
-        input.sharding_spec(),
-        input.size(),  # type: ignore[arg-type]
-        process_group=pg,
+_register_sharded_op_on_local_tensor(
+    torch.Tensor.masked_fill,
+    extra_check=sharded_masked_fill_check,
+    customized_func=sharded_masked_fill,
+)
+
+def sharded_view_check(*args, **kwargs):
+    """
+    Perform extra checks for the sharded_view op.
+
+    Args: same as ``torch.Tensor.view``.
+
+    Return: None
+    """
+    st = args[0]
+    shapes = args[1:]
+    if len(shapes) == 0:
+        raise ValueError("Missing *shape for sharded view op.")
+    if len(shapes) <= st.sharding_spec().dim:
+        raise NotImplementedError(
+            f"Shape having dim {len(shapes)} is not supported "
+            f"for sharded tensor sharded on dim {st.sharding_spec().dim}."
+        )
+    st_size = math.prod(st.size())
+    shape_size = math.prod(shapes)
+    if (
+        shape_size > st_size
+        or st_size % shape_size
+        or shapes.count(lambda x: x < -1) > 0
```

Comment:
Shouldn't this be <=1?

Reply:
Oh, one -1 is allowed, but more than one -1 is invalid.

Reply:
After a second thought on this, will simply this logic a little bit.